src/ripple/basics/XRPAmount.h

```diff
@@ -0,0 +1,388 @@
+        return *this;
+    }
+
+    XRPAmountBase&
+    operator-= (XRPAmountBase const& other)
+    {
+        drops_ -= other.drops();
+        return *this;
+    }
+
+    XRPAmountBase&
+    operator++()
+    {
+        ++drops_;
+        return *this;
+    }
+
+    XRPAmountBase&
+    operator--()
+    {
+        --drops_;
+        return *this;
+    }
+
+    XRPAmountBase&
+    operator*= (value_type const& rhs)
+    {
+        drops_ *= rhs;
+        return *this;
+    }
+
+    XRPAmountBase&
+    operator/= (value_type const& rhs)
+    {
+        drops_ /= rhs;
+        return *this;
+    }
+
+    XRPAmountBase&
+    operator%= (value_type const& rhs)
+    {
+        drops_ %= rhs;
+        return *this;
+    }
+
+    XRPAmountBase
+    operator- () const
+    {
+        static_assert( std::is_signed_v<T>,
+            "- operator illegal on unsigned XRPAmount types");
+        return { -drops_ };
+    }
+
+    bool
+    operator==(XRPAmountBase const& other) const
+    {
+        return drops_ == other.drops_;
+    }
+
+    bool
+    operator==(value_type other) const
+    {
+        return drops_ == other;
+    }
+
+    template <class Other,
+        class = enable_if_compatible_t <Other>>
+    bool
+    operator==(XRPAmountBase<Other> const& other) const
+    {
+        return drops_ == other.drops();
+    }
+
+    template <class Other,
+        class = enable_if_compatible_t <Other>>
+    bool
+    operator!=(XRPAmountBase<Other> const& other) const
+    {
+        return !operator==(other);
+    }
+
+    bool
+    operator<(XRPAmountBase const& other) const
+    {
+        return drops_ < other.drops_;
+    }
+
+    /** Returns true if the amount is not zero */
+    explicit
+    constexpr
+    operator bool() const noexcept
+    {
+        return drops_ != 0;
+    }
+
+    /** Return the sign of the amount */
+    constexpr
+    int
+    signum() const noexcept
+    {
+        return (drops_ < 0) ? -1 : (drops_ ? 1 : 0);
+    }
+
+    /** Returns the number of drops */
+    constexpr
+    value_type
+    drops () const
+    {
+        return drops_;
+    }
+
+    constexpr
+    double
+    decimalXRP () const;
+
+    template <class Dest,
+        class = std::enable_if_t<
+            std::is_same_v<typename Dest::unit_type, unit_type> &&
+            is_compatible_v <typename Dest::value_type> > >
+    Dest
+    as() const
+    {
+        using desttype = typename Dest::value_type;
+        if constexpr (is_safetocasttovalue_v <desttype, value_type>)
+        {
+            return { safe_cast<desttype>(drops_) };
+        }
+        else
+        {
+            if ((drops_ > std::numeric_limits<desttype>::max()) ||
+                (!std::numeric_limits<desttype>::is_signed && drops_ < 0) ||
+                (std::numeric_limits<desttype>::is_signed &&
+                    drops_ < std::numeric_limits<desttype>::lowest()))
+            {
+                Throw<std::runtime_error>("XRPAmount conversion out of range");
```

Comment:
I'd return an optional rather than throwing.

Reply:
@seelabs You're rewrite is a great simplification. I pushed your changes to the branch, and updated the unit test you removed.